src/functional/iterator/transforms/unroll_reduce.py

```diff
@@ -5,61 +5,79 @@
-        isinstance(arg, ir.FunCall)
-        and isinstance(arg.fun, ir.FunCall)
-        and arg.fun.fun == ir.SymRef(id="shift")
+        isinstance(arg, itir.FunCall)
+        and isinstance(arg.fun, itir.FunCall)
+        and arg.fun.fun == itir.SymRef(id="shift")
     )
 
 
-def _is_applied_lift(arg: ir.Expr) -> TypeGuard[ir.FunCall]:
+def _is_applied_lift(arg: itir.Expr) -> TypeGuard[itir.FunCall]:
     return (
-        isinstance(arg, ir.FunCall)
-        and isinstance(arg.fun, ir.FunCall)
-        and arg.fun.fun == ir.SymRef(id="lift")
+        isinstance(arg, itir.FunCall)
+        and isinstance(arg.fun, itir.FunCall)
+        and arg.fun.fun == itir.SymRef(id="lift")
     )
 
 
-def _is_shifted_or_lifted_and_shifted(arg: ir.Expr) -> TypeGuard[ir.FunCall]:
+def _is_shifted_or_lifted_and_shifted(arg: itir.Expr) -> TypeGuard[itir.FunCall]:
     return _is_shifted(arg) or (
         _is_applied_lift(arg)
         and any(_is_shifted_or_lifted_and_shifted(nested_arg) for nested_arg in arg.args)
     )
 
 
-def _get_shifted_args(reduce_args: Iterable[ir.Expr]) -> Iterator[ir.FunCall]:
+def _get_shifted_args(reduce_args: Iterable[itir.Expr]) -> Iterator[itir.FunCall]:
     return filter(
         _is_shifted_or_lifted_and_shifted,
         reduce_args,
     )
 
 
-def _is_list_of_funcalls(lst: list) -> TypeGuard[list[ir.FunCall]]:
-    return all(isinstance(f, ir.FunCall) for f in lst)
+def _is_list_of_funcalls(lst: list) -> TypeGuard[list[itir.FunCall]]:
+    return all(isinstance(f, itir.FunCall) for f in lst)
 
 
-def _get_partial_offset(arg: ir.FunCall) -> ir.OffsetLiteral:
+def _get_partial_offset_tag(arg: itir.FunCall) -> str:
     if _is_shifted(arg):
-        assert isinstance(arg.fun, ir.FunCall)
+        assert isinstance(arg.fun, itir.FunCall)
         offset = arg.fun.args[-1]
-        assert isinstance(offset, ir.OffsetLiteral)
-        return offset
+        assert isinstance(offset, itir.OffsetLiteral)
+        assert isinstance(offset.value, str)
+        return offset.value
     else:
         assert _is_applied_lift(arg)
         assert _is_list_of_funcalls(arg.args)
-        partial_offsets = [_get_partial_offset(arg) for arg in arg.args]
+        partial_offsets = [_get_partial_offset_tag(arg) for arg in arg.args]
         assert all(o == partial_offsets[0] for o in partial_offsets)
         return partial_offsets[0]
 
 
-def _get_connectivity(reduce_args: Iterable[ir.Expr], offset_provider) -> common.Connectivity:
-    connectivities = []
-    for arg in _get_shifted_args(reduce_args):
-        connectivities.append(offset_provider[_get_partial_offset(arg).value])
+def _get_partial_offset_tags(reduce_args: Iterable[itir.Expr]) -> Iterable[str]:
+    return [_get_partial_offset_tag(arg) for arg in _get_shifted_args(reduce_args)]
+
+
+def _is_reduce(node: itir.FunCall) -> TypeGuard[itir.FunCall]:
+    return isinstance(node.fun, itir.FunCall) and node.fun.fun == itir.SymRef(id="reduce")
+
+
+def get_connectivity(
```

Comment:
If this is public now, it feels like it doesn't belong into this pass anymore. Maybe a todo makes sense.

Reply:
made private again, this was because of intermediate steps...